datafusion/physical-plan/src/coop.rs

```diff
@@ -0,0 +1,325 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Utilities for improved cooperative scheduling.
+//!
+//! # Cooperative scheduling
+//!
+//! A single call to `poll_next` on a top-level `Stream` may potentially do a lot of work before it
+//! returns a `Poll::Pending`. Think for instance of calculating an aggregation over a large dataset.
+//! If an operator tree runs for a long period of time without yielding back to the Tokio executor,
+//! it can starve other tasks waiting on that executor to execute them.
+//! Additionally, this prevents the query execution from being cancelled.
+//!
+//! To ensure that `Stream` implementations yield regularly, operators can insert explicit yield
+//! points using the utilities in this module. For most operators this is **not** necessary. The
+//! built-in DataFusion operators that generate (rather than manipulate; for instance `DataSourceExec`)
+//! or repartition `RecordBatch`es (for instance, `RepartitionExec`) contain yield points that will
+//! make most operator trees yield as appropriate.
+//!
+//! There are a couple of types of operators that should insert yield points:
+//! - New source operators that do not make use of Tokio resources
+//! - Exchange like operators that do not use Tokio's `Channel` implementation to pass data between
+//!   tasks
+
+#[cfg(feature = "tokio_coop_fallback")]
+use futures::Future;
+use std::any::Any;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+use crate::execution_plan::CardinalityEffect::{self, Equal};
+use crate::{
+    DisplayAs, DisplayFormatType, ExecutionPlan, PlanProperties, RecordBatchStream,
+    SendableRecordBatchStream,
+};
+use arrow::record_batch::RecordBatch;
+use arrow_schema::Schema;
+use datafusion_common::{internal_err, Result, Statistics};
+use datafusion_execution::TaskContext;
+
+use crate::execution_plan::SchedulingType;
+use crate::stream::RecordBatchStreamAdapter;
+use futures::{Stream, StreamExt};
+
+/// A stream that passes record batches through unchanged while cooperating with the Tokio runtime.
+/// It consumes cooperative scheduling budget for each returned [`RecordBatch`],
+/// allowing other tasks to execute when the budget is exhausted.
```

Comment:
I think it would be useful to note the cooperating time is yielding back to the scheduler

```suggestion
/// It consumes cooperative scheduling budget for each returned [`RecordBatch`],
/// and returns control to the tokio scheduler periodically, allowing other tasks to 
/// execute or the plan to be cancelled when the budget is exhausted.
```